tools/bench.py

```diff
@@ -0,0 +1,220 @@
+################################################################################
+# Copyright (C) 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+# SPDX-License-Identifier: Apache-2.0
+################################################################################
+
+import ctypes
+import random
+import timeit
+from typing import List, Sequence, Tuple
+
+import numpy as np
+from gc_mlir import ir, runtime
+from gc_mlir.graph_compiler import GraphCompiler
+from utils import (
+    emit_benchmark_wrapped_main_func,
+    emit_nano_time,
+    get_kernel_func_from_module,
+)
+
+
+def py_timeit_bench(
+    ir_module: ir.Module,
+    entry_name: str,
+    pipeline: str,
+    mlir_args: list,
+    shared_libs: Sequence,
+    ir_printing=False,
+    repeat_time=100,
+    warm_up=20,
+) -> Tuple[float, float]:
+    """benchmark mlir with python timeit."""
+    compiler = GraphCompiler(
+        pipeline,
+        shared_libs,
+    )
+    compile_begin = timeit.default_timer()
+    engine = compiler.compile_and_jit(ir_module, ir_printing=ir_printing)
+    compile_cost = (timeit.default_timer() - compile_begin) * 1000
+
+    func = engine.lookup(entry_name)
+    packed_args = (ctypes.c_void_p * len(mlir_args))()
+    for argNum in range(len(mlir_args)):
+        packed_args[argNum] = ctypes.cast(mlir_args[argNum], ctypes.c_void_p)
+
+    def run_bench(func, arg):
+        func(arg)
+
+    timeit.timeit(lambda: run_bench(func, packed_args), number=warm_up)
+    total_time = timeit.timeit(lambda: run_bench(func, packed_args), number=repeat_time)
+    execute_cost = total_time * 1000 / repeat_time
+    return (execute_cost, compile_cost)
+
+
+def mlir_wrapper_bench(
+    ir_module: ir.Module,
+    entry_name: str,
+    pipeline: str,
+    mlir_args: list,
+    shared_libs: Sequence,
+    ir_printing=False,
+    repeat_time=100,
+    warm_up=20,
+) -> Tuple[float, float]:
+    """benchmark mlir with a wrapper func."""
+    kernel_func = get_kernel_func_from_module(ir_module, entry_name)
+    wrapper_module = ir_module
+    with ir.InsertionPoint(wrapper_module.body):
+        emit_benchmark_wrapped_main_func(kernel_func, emit_nano_time())
+    compiler = GraphCompiler(
+        pipeline,
+        shared_libs,
+    )
+    compile_begin = timeit.default_timer()
+    engine = compiler.compile_and_jit(wrapper_module, ir_printing=ir_printing)
+    compile_cost = (timeit.default_timer() - compile_begin) * 1000
```

Comment:
Can be a decorator

Reply:
do you mean add a decorator for the `compile_and_jit function`? if so, is there a way to get both `engine` and `compile_cost ` ?

Reply:
sry, I'm still confused about the decorator, is the follwing decorator something you expect?

```
def timing_decorator(func):
    def wrapper(*args, **kwargs):
        start_time = timeit.default_timer()
        result = func(*args, **kwargs)
        end_time = timeit.default_timer()  
        elapsed_time = (end_time - start_time) * 1000
        return elapsed_time, result
    return wrapper


# add decorator for compile_and_jit
@timing_decorator
def compile_and_jit(
    elf, module: ir.Module, ir_printing: bool = False
) -> execution_engine.ExecutionEngine:
    self.compile(module, ir_printing)
    return self.jit(module)


# invoke
compile_cost, engine = compiler.compile_and_jit(ir_module, ir_printing=ir_printing)

```